File: python/2021/05/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    A = u.read_match(file,"(\d+),(\d+) -> (\d+),(\d+)")
    return [[int(x) for x in a] for a in A]

def dump(N):
    for n in N:
        for x in n:
            print("%3s" % x, end="")
        print()
    print()

def sign(k):
    return 0 if k == 0 else k//abs(k)

def write(a,N):
    x,y,xx,yy = a[0],a[1],a[2],a[3]
    dx, dy = sign(xx-x), sign(yy-y)
    while True:
        N[y][x] += 1
        if x == xx and y == yy:
            break
        x += dx
        y += dy
    return N

def part1():
    A = read_input()
    maxX = max([a[0] for a in A]+[a[2] for a in A])+1
    maxY = max([a[1] for a in A]+[a[3] for a in A])+1
    A = [a for a in A if a[0]==a[2] or a[1]==a[3]]
    N = [[0 for i in range(maxX)] for j in range(maxY)]
    for a in A:
        N = write(a,N)
    k = sum([1 if x > 1 else 0 for n in N for x in n])
    print("Part 1:",k)

def part2():
    A = read_input()
    maxX = max([a[0] for a in A]+[a[2] for a in A])+1
    maxY = max([a[1] for a in A]+[a[3] for a in A])+1
    N = [[0 for i in range(maxX)] for j in range(maxY)]
    for a in A:
        N = write(a,N)
    k = sum([1 if x > 1 else 0 for n in N for x in n])
    print("Part 2:",k)

part1()
part2()

File: c/2021/05/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif

int16_t** getCoord(ARR_STRING* in) {
    int16_t** coord = malloc(in->length*sizeof(int16_t*));
    for (int i = 0; i < in->length; i++) {
        coord[i] = malloc(4*sizeof(int16_t));
        char* tmp;
        tmp = strtok(in->strings[i],", ->");
        int j = 0;
        while (tmp != NULL) {
            if (strlen(tmp) > 0) {
                coord[i][j++] = atoi(tmp);
            }
            tmp = strtok(NULL,", ->");
        }
    }
    return coord;
}

int16_t sign(int16_t x) {
    return x > 0 ? 1 : x < 0 ? -1 : 0;
}

void append(int64_t **map, int64_t* size, int64_t x, int64_t y) {
    for (int64_t i=0; i<(*size); i++) {
        if (map[i][0] == x && map[i][1] == y) {
            map[i][2] += 1;
            return;
        }
    }
    map[*size] = malloc(3*sizeof(int64_t));
    map[*size][0] = x;
    map[*size][1] = y;
    map[*size][2] = 1;
    (*size)++;
    return;
}

void printMap(int64_t **map, int64_t size) {
    for (int64_t i=0; i<size; i++) {
        printf("M %ld %ld %ld\n",map[i][0], map[i][1], map[i][2]);
    }
    printf("\n");
}

int64_t getCount(int16_t** coord, int64_t length, bool filter) {
    int64_t **map = malloc(500000*sizeof(int64_t*));
    int64_t size = 0;

    for (int64_t i = 0; i < length; i++) {
        if (filter && (coord[i][0] != coord[i][2] && coord[i][1] != coord[i][3])) continue;
        int64_t xD = coord[i][2]-coord[i][0];
        int64_t yD = coord[i][3]-coord[i][1];
        int64_t xS = sign(xD);
        int64_t yS = sign(yD);
        int64_t x,y;
        for (x=0,y=0; x!=xD || y!=yD; x+=xS,y+=yS) {
            append(map, &size, coord[i][0]+x, coord[i][1]+y);
        }
        append(map, &size, coord[i][0]+x, coord[i][1]+y);
    }
    
    int64_t count = 0;
    for (int64_t i = 0; i < size; i++) {
        if (map[i][2] > 1) count++;
    }
    for (int64_t i = 0; i<size; i++) {
        free(map[i]);
    }
    free(map);
    return count;
}

void part1() {
    ARR_STRING* in = readInput(FILE_NAME);
    int16_t** coord = getCoord(in);
    int64_t k = getCount(coord,in->length,true);
    printf("Part 1: %ld\n", k);
    for (int i = 0; i < in->length; i++) {
        free(coord[i]);
    }
    free(coord);
    freeArrString(in);
}

void part2() {
    ARR_STRING* in = readInput(FILE_NAME);
    int16_t** coord = getCoord(in);
    int64_t k = getCount(coord,in->length,false);
    printf("Part 2: %ld\n", k);
    for (int i = 0; i < in->length; i++) {
        free(coord[i]);
    }
    free(coord);
    freeArrString(in);
}

int main(void) {
    printf("This solution takes long time to execute, but only up to a minute.\n");
    part1();
    part2();
    return 0;
}
